### ISSUE TITLE
Serialization into incorrect OWL syntax

### ISSUE THREAD

**Issue by aifargonos** [owlcs/owlapi#468]
Yet another case when owlapi outputs file that is not accepted by the grammar of OWL Functional Syntax is when `ObjectIntersectionOf` has two arguments that are syntactically equivalent, for example:
```
Prefix(:=<http://www.example.com/test#>)
Ontology(<http://www.example.com/test>

	SubClassOf(
		:C698
		ObjectIntersectionOf(
			ObjectSomeValuesFrom( :R125 ObjectIntersectionOf( :C691 :C690 ) )
			ObjectSomeValuesFrom( :R125 ObjectIntersectionOf( :C691 :C690 ) )
		)
	)

)
```
is serialized to
```
Prefix(:=<http://www.example.com/test#>)
Prefix(owl:=<http://www.w3.org/2002/07/owl#>)
Prefix(rdf:=<http://www.w3.org/1999/02/22-rdf-syntax-ns#>)
Prefix(xml:=<http://www.w3.org/XML/1998/namespace>)
Prefix(xsd:=<http://www.w3.org/2001/XMLSchema#>)
Prefix(rdfs:=<http://www.w3.org/2000/01/rdf-schema#>)


Ontology(<http://www.example.com/test>

Declaration(Class(:C690))
Declaration(Class(:C691))
Declaration(Class(:C698))
Declaration(ObjectProperty(:R125))

############################
#   Classes
############################

# Class: :C698 (:C698)

SubClassOf(:C698 ObjectIntersectionOf(ObjectSomeValuesFrom(:R125 ObjectIntersectionOf(:C690 :C691))))


)
```
which cannot be loaded by owlapi.

I suppose this is happening for all the constructs that should take at least two arguments and are backed by sets.

I tried this with version 3.5.0 and version 4.1.3 and it works wrong in both.

I think the easiest solution would be to adapt serializers so that they output the same argument twice, when the argument set contains only one element.

This is related to #151 and #46 (and some other issues), but it seems like each time there was only small fix of a specific problem, but not much was done about the common reason.

**Comment by ignazio1977** [comment]
The common reason being that that's an invalid axiom according to the specs. (But that's a rant I've had often, sorry about that).

The parser is supposed to suspend disbelief when receiving such an axiom - to be precise, the owldatafactory should take care of that and duplicate the value. On output, the axiom should show duplication or be dropped - at least I expect it in 4.1.3. In 3.5.0 this might have still been broken, can't recall.

**Comment by aifargonos** [comment]
The axiom does not have duplicated argument. The intersection has duplicated argument.

I saw that this was dealt with for axioms, but here it is the class expression that is causing problems.

**Comment by ykazakov** [comment]
Ignazio,

I guess you refer to the point 2.1 of the OWL 2 structural specification

http://www.w3.org/TR/owl2-syntax/#Structural_Specification

in which it is written: "By default, all associations are sets; that is, the objects in them are unordered and repetitions are disallowed." Well, then this slightly modified example should still be legit, is it?

SubClassOf(
        :C698
        ObjectIntersectionOf(
            ObjectSomeValuesFrom( :R125 ObjectIntersectionOf( :C691 :C690 ) )
            ObjectSomeValuesFrom( :R125 ObjectIntersectionOf( :C690 :C691 ) )
        )
    )

In principle, there is no repetition in ObjectIntersectionOf (the conjuncts are structurally equivalent, but not equal). I guess this will trigger the same problem in OWL API. Or you think in the OWL 2 specs repetitions should also include repetitions modulo structural equivalence?

**Comment by ignazio1977** [comment]
The way I understood it, "unordered and repetitions disallowed" makes the intersections of C691 and C690 indistinguishable - in other words, the OWLAPI represents them as .equals().

Semantically, the two are not different in any way, I believe - so in what context is it meaningful to consider them distinct? Honest question because I'm not sure I'm aware of all implications.

To fix this in the general case, we'd have all these be lists, minus duplication I assume. It would be a profound change, and I'm sure it would cause confusion for users of the library.

**Comment by ignazio1977** [comment]
I've misinterpreted the first report - yes expressions with only one element are a different case than, say, equivalent axioms, although I see the same issue in them not being valid OWL.

**Comment by aifargonos** [comment]
My original concern was about OWL exchange syntaxes. The mentioned section of the specification says:
> Sets written in one of the exchange syntaxes (e.g., XML or RDF/XML) are not necessarily expected to be duplicate free.
So according to my understanding the first example I gave is a valid OWL Functional Syntax, but the second example is not valid OWL Functional Syntax. Now I am not concerned about whether the created objects are valid OWL or not. The problem is just that owlapi creates a file that is supposed to be a valid OWL Functional-Style Syntax, but it is not.

I understand that changing all the sets to list is too big change, but what about adapting only the OWLStorers, so that they output a single argument two times?

**Comment by ignazio1977** [comment]
That's a good workaround, but I'm worried more about what the meaning of A and A is understood to be. Looks like it should just be rewritten as A; however it's a separate concern.

Is it OK for this to be fixed in 3.5.5 and 4.1.4, or are you tied to versions used in other tools?

**Comment by aifargonos** [comment]
It is ok if it is fixed in 3.5.5 and 4.1.4, thanks!
